source/s3_meta_request.c

```diff
@@ -842,20 +832,269 @@ static void s_s3_meta_request_send_request_finish_default(
+    AWS_ASSERT(message_headers);
+
+    struct aws_byte_cursor host_header_value;
+
+    bool host_header_exists =
+        aws_http_headers_get(message_headers, host_header_name, &host_header_value) == AWS_ERROR_SUCCESS;
+    AWS_ASSERT(host_header_exists);
+    (void)host_header_exists;
+
+    /* Try to acquire a token so that we can schedule a retry. */
+    if (aws_retry_strategy_acquire_retry_token(
+            retry_strategy, &host_header_value, s_s3_meta_request_queue_retry_with_token, request, 0)) {
+
+        AWS_LOGF_ERROR(
+            AWS_LS_S3_META_REQUEST,
+            "id=%p Meta request could not acquire retry token for request %p due to error %d (%s)",
+            (void *)request->meta_request,
+            (void *)request,
+            aws_last_error(),
+            aws_error_str(aws_last_error()));
+
+        goto error_ref_clean_up;
+    }
+
+    return;
+
+error_ref_clean_up:
+
+    aws_s3_request_release(request);
+    aws_s3_meta_request_finish(meta_request, NULL, 0, aws_last_error());
+    aws_s3_meta_request_release(meta_request);
+}
+
+void aws_s3_meta_request_handle_error(
+    struct aws_s3_meta_request *meta_request,
+    struct aws_s3_request *request,
+    int error_code) {
+    AWS_PRECONDITION(meta_request);
+
+    AWS_LOGF_ERROR(
+        AWS_LS_S3_META_REQUEST,
+        "id=%p Meta request handling error %d (%s). (request=%p)",
+        (void *)meta_request,
+        error_code,
+        aws_error_str(error_code),
+        (void *)request);
+
+    int response_status = 0;
+
+    if (request != NULL) {
+        request->send_data.error_code = error_code;
+        response_status = request->send_data.response_status;
+    }
+
+    /* If the request is NULL or we have a response that indicates retrying would be futile, then bail on the entire
+     * meta request. */
+    if (request == NULL || error_code == AWS_ERROR_S3_INVALID_RESPONSE_STATUS) {
+
+        AWS_LOGF_ERROR(
+            AWS_LS_S3_META_REQUEST,
+            "id=%p Meta request cannot recover from error %d (%s). (request=%p, response status=%d)",
+            (void *)meta_request,
+            error_code,
+            aws_error_str(error_code),
+            (void *)request,
+            response_status);
+
+        aws_s3_meta_request_finish(meta_request, request, response_status, error_code);
+        return;
+    }
+
+    struct aws_s3_client *client = aws_s3_meta_request_acquire_client(meta_request);
+
+    /* If we were able to get the point of handling an error, the client should still be around. */
+    AWS_ASSERT(client);
+    AWS_ASSERT(request);
+    AWS_ASSERT(request->meta_request == meta_request);
+
+    aws_s3_request_acquire(request);
+    aws_s3_meta_request_acquire(meta_request);
+
+    /* If the retry token is NULL, try to grab one, otherwise, just re-use the token from the request's last retry. */
+    if (request->retry_token == NULL) {
+        s_s3_meta_request_acquire_retry_token(client->retry_strategy, request);
```

Comment:
Can either of these fail?

Reply:
Internally they handle any failures right now.  We might be able to factor that out to this level if that's preferable.